core/domain/exp_domain.py

```diff
@@ -3519,3 +3519,226 @@ def to_metadata_dict(self):
+
+    def __init__(
+            self, exploration_id, exploration_version, state_name_to_ids,
+            latest_state_id_used):
+        """Initialize state id mapping model object.
+
+        Args:
+            exploration_id: str. The exploration id whose state names are
+            mapped.
+            exploration_version: int. The version of the exploration.
+            state_name_to_ids: dict. A dictionary mapping a state name to
+                unique ID.
+            latest_state_id_used: int. Last used state id.
+        """
+        self.exploration_id = exploration_id
+        self.exploration_version = exploration_version
+        self.state_name_to_ids = state_name_to_ids
+        self.latest_state_id_used = latest_state_id_used
+
+    @staticmethod
+    def has_state_changed_significantly(old_state, new_state):
+        """Checks whether there is any significant change in old and new version
+        of the given state.
+
+        A siginificant change has happend when interaction id has been changed
+        in new version of the state.
+
+        Args:
+            old_state: dict. Old version of the state.
+            new_state: dict. New version of the state.
+
+        Returns:
+            bool. A boolean indicating whether change is significant or not.
+        """
+        if old_state.interaction.id != new_state.interaction.id:
+            # State has changed significantly is interaction id in new state is
+            # different from old state.
+            return True
+
+    @classmethod
+    def get_state_id_mapping(
+            cls, old_exploration, new_exploration, change_list):
+        """Get state name to id mapping for new exploration's states.
+
+        Args:
+            old_exploration: Exploration. Old version of the exploration.
+            new_exploration: Exploration. New version of the exploration.
+            change_list: list(dict). List of changes between old and new
+                exploration.
+
+        Returns:
+            StateIdMapping. Domain object for state id mapping.
+        """
+        new_state_names = []
+        state_id_to_names = {}
+
+        # Get state id mapping model for old exploration.
+        model = cls.get_state_id_mapping_from_model(
+            exp_models.StateIdMappingModel.get_state_id_mapping_model(
+                old_exploration.id, old_exploration.version, strict=True))
+
+        # Generate state id to name mapping dict for old exploration.
+        for state_name in old_exploration.states:
+            state_id_to_names[model.get_state_id(state_name)] = state_name
+
+        old_state_id_to_names = copy.deepcopy(state_id_to_names)
+
+        # Analyse each command in change list one by one to create state id
+        # mapping for new exploration.
+        for change_dict in change_list:
+            if change_dict['cmd'] == CMD_ADD_STATE:
+                new_state_names.append(change_dict['state_name'])
```

Comment:
As a general note, you might as well add some assertions throughout that double-check that all assumptions hold, e.g. do "assert change_dict['state_name'] not in state_ids_to_names.values()" here. If any errors arise during a test run then this will help us debug. I think the production data should be clean, but you can't be too careful.

Reply:
Added some asserts, here and below.